### ISSUE TITLE
Wrong semantics in omelasticsearch createMsgFromRequest()

### ISSUE THREAD

**Issue by snaix** [rsyslog/rsyslog#3573]
### Expected behavior
As createMsgFromRequest() comment said in plugins/omelasticsearch.c :

>request string looks like this:
>"{\"create\":{\"_index\": \"rsyslog_testbench\",\"_type\":\"test-type\",
>\"_id\":\"FAEAFC0D17C847DA8BD6F47BC5B3800A\"}}\n
>{\"msgnum\":\"x00000000\",\"viaq_msg_id\":\"FAEAFC0D17C847DA8BD6F47BC5B3800A\"}\n"
>**we don't want the meta header, only the data part**
>start = first \n + 1
>end = last \n 

An expected behavior is rsyslog can get rawmsg as data part.

### Actual behavior
Only message format that contains "message" filed can be extracted from request, but not for custom message format.
If there is no message field, all request text will be filled for rawmsg.
It's confusing when use omelasticsearch retry. user can not get a correct rawmsg from rsyslog, and could lead a data lost because a wrong json message(contain meta header) would be sent to ES.

### Environment
The latest code shows that this problem still exists.

**Comment by davidelang** [comment]
can you please show a message written with the RSYSLOG_DebugFormat that has the 
problem you are describing?

I am not understanding the issue.

David Lang

**Comment by snaix** [comment]
> can you please show a message written with the RSYSLOG_DebugFormat that has the problem you are describing? I am not understanding the issue. David Lang

For instance:
I have a ruleset A for a normal omelasticsearch action, and another ruleset B for retryRuleset when  omelasticsearch failed for 429 status .
Obviously, the original, failed message  from ruleset A is needed in ruleset B .
Current, ruleset B extract the original message from elasticsearch reqeuset like:

> "{"create":{"_index": "rsyslog_testbench","_type":"test-type",
> "_id":"FAEAFC0D17C847DA8BD6F47BC5B3800A"}}\n
> {"msgnum":"x00000000","viaq_msg_id":"FAEAFC0D17C847DA8BD6F47BC5B3800A"}\n"

The original message is the data part, we needed:
> {"msgnum":"x00000000","viaq_msg_id":"FAEAFC0D17C847DA8BD6F47BC5B3800A"}\n"

But only data part contains "message" field can be extracted, or omelasticsearch will keep the whole request in rawmsg.
so we get a rawmsg like this:
> "{"create":{"_index": "rsyslog_testbench","_type":"test-type",
> "_id":"FAEAFC0D17C847DA8BD6F47BC5B3800A"}}\n
> {"msgnum":"x00000000","viaq_msg_id":"FAEAFC0D17C847DA8BD6F47BC5B3800A"}\n"

but not:
> {"msgnum":"x00000000","viaq_msg_id":"FAEAFC0D17C847DA8BD6F47BC5B3800A"}\n"

**Comment by rgerhards** [comment]
I admit I do not fully understand the question. @richm I guess you can shed some light?

**Comment by richm** [comment]
The problem is here:
```
	if (fjson_object_object_get_ex(jo_request, "message", &jo_msg)) {
		const char *rawmsg = json_object_get_string(jo_msg);
		const size_t msgLen = (size_t)json_object_get_string_len(jo_msg);
		MsgSetRawMsg(*msg, rawmsg, msgLen);
	} else {
		MsgSetRawMsg(*msg, request, strlen(request));
	}
```
If the original request has a field `"message"` we use the value of that field as the `$rawmsg` value for the record.  If it did not have  `"message"`, we just use the entire request - the metadata + the data.  I'm not sure, but I'm guessing the presence of the extra newline `\n` in the rawmsg causes problems when the data is resent to Elasticsearch?  Are you sending the rawmsg field to Elasticsearch?

@snaix is asking that instead, the rawmsg should be only the data part, not the metadata, so something like this:
```
	if (fjson_object_object_get_ex(jo_request, "message", &jo_msg)) {
		const char *rawmsg = json_object_get_string(jo_msg);
		const size_t msgLen = (size_t)json_object_get_string_len(jo_msg);
		MsgSetRawMsg(*msg, rawmsg, msgLen);
	} else {
		MsgSetRawMsg(*msg, datastart, datalen);
	}
```

**Comment by richm** [comment]
@snaix if you are able to run from source, please try https://github.com/rsyslog/rsyslog/pull/3577

**Comment by rgerhards** [comment]
closed via #3577

**Comment by snaix** [comment]
@richm Yes, that's what I want to say. I want to get the data part but not entire raw message (meta + data), because I can't parse it.
Thank you for your excellent work!
